File: Searching Techniques/Binary Search/Binary_Search.cpp
#include <iostream>
using namespace std;

int binarySearch(int arr[], int n, int key)
{
    int start = 0;
    int end = n - 1;

    int mid = (start + end) / 2;
    while (start <= end)
    {
        if (arr[mid] == key)
        {
            return mid;
        }
        if (key > arr[mid])
        {
            start = mid + 1;
        }
        else
        {
            end = mid - 1;
        }
        mid = (start + end) / 2;
    }
    return -1;
}

int main()
{
    int size;
    cout << "Enter Size of Array:";
    cin >> size;
    int arr[size];
    cout << "Enter Array Elements:";
    for (int i = 0; i < size; i++)
    {
        cin >> arr[i];
    }
    int key;
    cout << "Enter Key to Search:";
    cin >> key;
    int index = binarySearch(arr, size, key);
    cout << "Index of " << key << " is:" << index << endl;
    return 0;
}

File: Recursion/BinarySearch.java

public class BinarySearch {

    public static int search(int start, int end, int target, int[] arr) {
        if (start > end) {
            return -1;
        }
        int mid = start + (end - start) / 2;

        if (target == arr[mid]) {
            return mid;
        }

        if (target < arr[mid]) {
            return search(start, mid - 1, target, arr);
        } else {
            return search(mid + 1, end, target, arr);
        }
    }

    public static void main(String[] args) {
        int[] arr = {1, 2, 3, 4, 5, 6};
        int start = 0;
        int end = arr.length - 1;
        int target = 4;
        int ans = search(start, end, target, arr);
        System.err.println(ans);
    }
}
